Alternative implementation

rhaschke opened the pull request:
Implement the desired behavior by intercepting `add_argument()` and installing additional type resolvers for folder arguments to be resolved w.r.t. the workspace root.
Realizes https://github.com/colcon/colcon-core/pull/595#issuecomment-1800424329
Fixes https://github.com/colcon/colcon-top-level-workspace/pull/1#issuecomment-1420033687

rhaschke committed:
Implementation using argument_parser

rhaschke committed:
Use log in workspace root

The --log-base argument is handled differently than other base-folder arguments.
Its default value is None, meaning that the environment variable COLCON_LOG_PATH
or the default path is used if the command-line argument is missing.
As we cannot easily change the default path (which is hardcoded in colcon-core),
we override the environment variable instead.

rhaschke committed:
Resolve "ROOT" in --base-paths argument

rhaschke committed:
Use logical working directory, not resolving symlinks

When running from a symlinked subfolder, os.getcwd() and Path.cwd()
return the real path with all all symlinks resolved.
This resulted in the workspace root to be found w.r.t. the real path
instead the actual logical one.

rhaschke committed:
Improve README

Timple commented:
Awesome, thank you for implementing symlink resolving. We use those as well :slightly_smiling_face:

rhaschke commented:
Ping @cottsay for a review of this.

andrejpan commented:
How exactly could I use this? I glanced at Colcon docs. However, I did find out how I could install the extension manually.

agyoungs commented:
This looks to be pretty useful. Is there any movement on this?

yurirocha15 commented:
> How exactly could I use this? I glanced at Colcon docs. However, I did find out how I could install the extension manually.

Just in case anyone else has the same question: You can do
```bash
python3 -m pip install git+https://github.com/rhaschke/colcon-top-level-workspace.git
```
to install it (as long colcon is using the same python interpreter).

ArendJan committed:
Print used root path to stderr (#2)

This way, this informational output doesn't interfere with the desired output from colcon itself.

rhaschke committed:
0.1.0

rhaschke committed:
Silent flake8 warnings

rhaschke commented:
I released this to pypi now. Unfortunately, there was no activity here for years.
